src/librustc_typeck/error_codes.rs

```diff
@@ -2425,6 +2425,83 @@ struct Bar<S, T> { x: Foo<S, T> }
 ```
 "##,
 
+E0307: r##"
+This error indicates that the `self` parameter in a method has an invalid
+"reciever type".
+
+Methods take a special first parameter, of which there are three variants:
+`self`, `&self`, and `&mut self`. The type `Self` acts as an alias to the
+type of the current trait implementor, or "receiver type". Besides the
+already mentioned `Self`, `&Self` and `&mut Self` valid receiver types, the
+following are also valid, if less common: `self: Box<Self>`,
+`self: Rc<Self>`, `self: Arc<Self>`, and `self: Pin<P>` (where P is one of
+the previous types except `Self`).
+
+```
+# struct Foo;
+trait Trait {
+    fn foo(&self);
+//         ^^^^^ this let's you refer to the type that implements this trait
```

Comment:
No it doesn't; `self` lives only in the value namespace. Only `Self` is the type. (See above re. explaining `self` as sugar)

Reply:
What would be an appropriate one sentence explanation on what `&self` is in the trait?

Reply:
"`self` here is a reference to the receiver object" maybe?